sumpy/recurrence.py

```diff
@@ -0,0 +1,607 @@
+    terms = np.array(terms)
+
+    # Sort terms and create idx_l
+    idx_l = []
+    for i in range(len(terms)):
+        tms = list(terms[i].atoms(sp.Number))
+        if len(tms) == 1:
+            idx_l.append(tms[0])
+        else:
+            idx_l.append(0)
+    idx_l = np.array(idx_l, dtype='int')
+    idx_sort = idx_l.argsort()
+    idx_l = idx_l[idx_sort]
+    terms = terms[idx_sort]
+
+    # Order is the max difference between highest/lowest in idx_l
+    order = max(idx_l) - min(idx_l) + 1
+
+    # How much do we need to shift the recurrence relation
+    shift_idx = max(idx_l)
+
+    # Get the respective coefficients in the recurrence relation from r
+    n = sp.symbols("n")
+    s = sp.Function("s")
+    coeffs = sp.poly(r, list(terms)).coeffs()
+
+    # Re-arrange the recurrence relation so we get s(n) = ____
+    # in terms of s(n-1), ...
+    true_recurrence = sum([coeffs[i]/coeffs[-1] * terms[i]
+                           for i in range(0, len(terms)-1)])
+    true_recurrence1 = true_recurrence.subs(n, n-shift_idx)
+
+    if replace:
+        # Replace s(n-1) with snm_1, s(n-2) with snm_2 etc.
+        # because pymbolic.substitute won't recognize it
+        last_syms = [sp.Symbol(f"anm{i+1}") for i in range(order-1)]
+        # pylint: disable=not-callable
+        # Assumes order > 1
+        true_recurrence2 = true_recurrence1.subs(s(n-1), last_syms[0])
+        for i in range(2, order):
+            true_recurrence2 = true_recurrence2.subs(s(n-i), last_syms[i-1])
+        return order, true_recurrence2
+
+    return order, true_recurrence1
+
+
+def extract_idx_terms_from_recurrence(r: sp.Expr) -> tuple[np.ndarray,
+                                                           np.ndarray]:
+    r"""
+    Given a recurrence extracts the variables in the recurrence
+    as well as the indexes in sorted order.
+
+    :arg r: recurrence to extract terms from
+    """
+    terms = list(r.atoms(sp.Function))
+    terms = np.array(terms)
+
+
+    idx_l = []
+    for i in range(len(terms)):
+        tms = list(terms[i].atoms(sp.Number))
+        if len(tms) == 1:
+            idx_l.append(tms[0])
+        else:
+            idx_l.append(0)
+    idx_l = np.array(idx_l, dtype='int')
+    idx_sort = idx_l.argsort()
+    idx_l = idx_l[idx_sort]
+    terms = terms[idx_sort]
+
+    return idx_l, terms
+
+
+def __check_neg_ind(r_n):
+    r"""
+    Simply checks if a negative index exists in a recurrence relation.
+    """
+
+    idx_l, _ = extract_idx_terms_from_recurrence(r_n)
+
+    return np.any(idx_l < 0)
+
+
+def __get_initial_c(recurrence):
+    r"""
+    For a given recurrence checks how many initial conditions by
+    checking for non-negative indexed terms.
+    """
+    n = sp.symbols("n")
+
+    i = 0
+    r_c = recurrence.subs(n, i)
+    while __check_neg_ind(r_c):
+        i += 1
+        r_c = recurrence.subs(n, i)
+    return i
+
+
+def shift_recurrence(r: sp.Expr, var: np.ndarray) -> sp.Expr:
```

Comment:
Can we generate those in the first place, to avoid having to have two "colors" of recurrence?